### PR TITLE
GeneTek tgui

### PR DESCRIPTION
**Pull request by BenLubar**
[FEATURE]

## About the PR

- Reimplements the GeneTek scanner UI in tgui.
- Removes the concept of "tracking" research. The timers for all research operations are now automatically shown.
- Non-scientific observation: Every action takes about half as many clicks now.
- Makes NPC monkeys with AI have randomized appearances/ages/blood types. Non-AI NPC monkeys already had this.
- Adds an admin verb "See No Evil" to blindfold all monkeys. (Mostly because I got tired of doing this manually while testing.)
- When entering the scanner, the interior monitor now powers on automatically rather than requiring the occupant to be staring at the console beforehand.
- Using the radiation emitter now keeps the mutant race of the occupant if it would have been possible to reapply using the Body menu.
- Entering the gene scanner with an active version of a researched gene sets the gene to "activated" status.
- Added a bunch of fun sound effects to the GeneTek scanner console. (eg. when printing an injector or using the radiation emitter)

Note: This does not change the Appearance menu as this menu is very rarely used.

## Changelog

```
(u)BenLubar:
(*)GeneTek has released the next generation of their gene scanner computer interface!
(+)Lab monkeys now have randomized ages and blood types.
```

### CODE REVIEW

**Comment by BenLubar** [comment]
Comparison:

image

**Comment by BenLubar** [comment]
At this point, I'm not sure if I should spend the effort to change the appearance editor, since I'm not sure anyone uses it.

If anyone wants to mess around with the UI and let me know if there are any problems, that'd be great.

I'd like to wait until #3213 is merged (and then fix the resulting merge conflict) as this PR contains an older version of that PR before marking this as non-draft.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/GeneTek.js:891]: `-10px` margin seems pretty interesting. What's this for?

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/GeneTek.js:962]: Prefer a mapping of number -> object with relevant properties for this (like the occupant status in the Sleeper).

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/GeneTek.js:1255]: What is the significance of `"5"`?

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
I've not looked at any of the `geneticsMachines.dm` changes yet.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (BenLubar) [reply tgui/packages/tgui/interfaces/GeneTek.js:1255]: The number ranges from 1 to 5, with 1 as green, 3 as yellow, and 5 as red.

**Comment by mordent-goonstation** [comment]
> At this point, I'm not sure if I should spend the effort to change the appearance editor, since I'm not sure anyone uses it.
> 
> If anyone wants to mess around with the UI and let me know if there are any problems, that'd be great.

It's definitely used. I would look to see about a way to add character appearance editing to this PR (with a mind to it being re-used in the character setup page at some point) prior to it being merged.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (mordent-goonstation) [reply tgui/packages/tgui/interfaces/GeneTek.js:1255]: Rephrasing: this info should be added to the code.

File: tgui/packages/tgui/styles/components/Dimmer.scss
Comment (BenLubar) [reply tgui/packages/tgui/styles/components/Dimmer.scss:13]: This makes it so scrollable windows can have modals in them.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (BenLubar) [reply tgui/packages/tgui/interfaces/GeneTek.js:247]: I think this one is a remnant of a previous way I was laying out the knob.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (BenLubar) [reply tgui/packages/tgui/interfaces/GeneTek.js:423]: `tier` is the index on the byond side (with 1-based indexing), so `finishedResearch[i]` in JavaScript ranges from `i=0` to `i=3`

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (BenLubar) [reply tgui/packages/tgui/interfaces/GeneTek.js:734]: The server sends the occupant icon using `browse_rsc`, and the icon is rebuilt whenever someone enters the scanner or something the scanner does forces a rebuild (such as changing mutant race). Byond docs say this means the icon is only sent when it's changed.

**Comment by BenLubar** [comment]
Here are the modals:

image
image

**Comment by BenLubar** [comment]
Here's the WIP appearance editor

image
image

File: code/modules/medical/genetics/geneticsMachines.dm
Comment (ZeWaka) [inline code/modules/medical/genetics/geneticsMachines.dm:26]: FALSE?

File: code/modules/medical/genetics/geneticsMachines.dm
Comment (ZeWaka) [inline code/modules/medical/genetics/geneticsMachines.dm:274]: Note: behavior change

**Review by ZeWaka — COMMENTED** [review]
DM looks good, looked over all but the JS blobs

File: code/modules/medical/genetics/geneticsMachines.dm
Comment (BenLubar) [reply code/modules/medical/genetics/geneticsMachines.dm:26]: I believe these are "timestamp when next available" but I'd have to double check.

File: code/modules/medical/genetics/geneticsMachines.dm
Comment (BenLubar) [reply code/modules/medical/genetics/geneticsMachines.dm:274]: This one actually isn't a behavior change. You were previously able to watch (but not interact with) the console while inside it but only if you opened the console before going into the pod.

**Comment by optimumtact** [comment]
sexy use of dials

File: code/modules/medical/genetics/geneticsScanner.dm
Comment (mordent-goonstation) [inline code/modules/medical/genetics/geneticsScanner.dm:278]: Sanity check on how often this line gets called. This uses `getFlatIcon` under the hood, which isn't particularly cheap. AFAICT `update_occupant` is called whenever the `editappearance` action is sent, which seems to be done whenever any of those controls are messed with. Big concern for me is dragging those dials around slowly causing *lots* of calls here.

Comment (BenLubar) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: The dials only send the signal back to the server when you release them. Not really much more efficient I can make it unless I render the character sprite on the client side somehow.

Comment (pali6) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: No way, current character setup lags the server a lot when someone decides to edit their profiles during the round. Replace this with something else.

Comment (mordent-goonstation) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: @BenLubar I'm going to have to dabble with working out how to show changing sprites (for this and for an inevitable rework of the Character Setup popup) efficiently should work. I am aware that there's a "ByondUi" component that might be useful here, you're welcome to take a look into that if you'd like, else I'll add to my to-do list.

If you do look into it, please try to make something we can re-use for showing a character sprite rather than baking into the genetics console stuff.

Comment (optimumtact) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: Have you considered changing the character preview a map window instead? we did it on /tg/ and it greatly reduce lag from character setup

Comment (BenLubar) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: If someone can do that, be my guest!

I have no idea how the map window stuff works in byond after looking at it quite a bit.

Comment (ZeWaka) [reply code/modules/medical/genetics/geneticsScanner.dm:278]: > If someone can do that, be my guest!
> 
> I have no idea how the map window stuff works in byond after looking at it quite a bit.

TG's character setup has an example.

File: tgui/packages/tgui/interfaces/GeneTek.js
Comment (BenLubar) [reply tgui/packages/tgui/interfaces/GeneTek.js]: It makes the modal not have a huge amount of padding around the section. Should probably move that into a stylesheet so everyone gets the benefit.

**Comment by ZeWaka** [comment]
[Ready for review]?

**Comment by BenLubar** [comment]
> [Ready for review]?

This contains a slightly outdated version of #3360, but it's just code formatting changes that are missing.

**Comment by ZeWaka** [comment]
one last thing in mordent's review

**Comment by BenLubar** [comment]
> one last thing in mordent's review

Ah! I keep seeing that, going "that'll be easy to fix", and then forgetting because GitHub likes to hide that comment for no reason.

**Pull request merged**
